File: kombi-app/src/main/java/com/sgnatiuk/combination/CombinationUsage.java
package com.sgnatiuk.combination;

import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.stream.IntStream;

import static java.util.stream.Collectors.toMap;

public class CombinationUsage {

    public static void main(String[] args) {
        List<String> inputData = Arrays.asList("A", "B", "C");
        printCartesianProductFromList(inputData);
        printCartesianProductFromMap(inputData);
    }

    public static <T> void printCartesianProductFromList(List<T> inputData) {
        System.out.println("Combinations of " + inputData);
        CombinationsBuilder.combinationsOf(inputData)
                .forEach(System.out::println);
    }

    public static <T> void printCartesianProductFromMap(List<T> inputData) {
        Map<Integer, T> inputDataMap = IntStream.range(0, inputData.size())
                .boxed().collect(
                        toMap(i -> i, inputData::get)
                );
        System.out.println("Combinations of " + inputDataMap);
        CombinationsBuilder.combinationsOf(inputDataMap)
                .forEach(System.out::println);
    }
}

File: kombi-app/src/main/kotlin/com/sgnatiuk/combination/CombinationUsage.kt
package com.sgnatiuk.combination

import com.sgnatiuk.listOfStrings
import com.sgnatiuk.mapIntToString

fun main(args: Array<String>) {
    printCombinationsOfList(listOfStrings)
    printCombinationsOfMap(mapIntToString)
}

fun <T> printCombinationsOfList(inputData: List<T>) {
    println("Combinations of $inputData")
    combinationsOf(inputData).forEach(::println)
}

fun <K,V> printCombinationsOfMap(inputData: Map<K, V>) {
    println("Combinations of $inputData")
    combinationsOf(inputData).forEach(::println)
}